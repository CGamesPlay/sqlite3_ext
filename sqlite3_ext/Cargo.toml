[package]
name = "sqlite3_ext"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[features]
static = [ "dep:rusqlite", "dep:libsqlite3-sys" ]
static_modern = [ "static", "rusqlite?/modern_sqlite" ]

[dependencies]
once_cell = "1.0"
sqlite3_ext_macro = { path = "sqlite3_ext_macro" }
rusqlite = { version = "0.27.0", optional = true }
libsqlite3-sys = { version = "0.24.0", optional = true }

[build-dependencies]
proc-macro2 = "1.0"
which = "4.2.5"
syn = { version = "1.0", features = [ "parsing", "full" ] }
quote = "1.0"

[[test]]
name = "read_only"
required-features = [ "static" ]

[[test]]
name = "update"
required-features = [ "static" ]

[[example]]
name = "generate_series"
crate-type = [ "staticlib", "cdylib" ]
test = true
