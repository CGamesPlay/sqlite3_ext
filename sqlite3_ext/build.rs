use proc_macro2::{Ident, Span, TokenStream};
use quote::*;
use std::env;
use std::fs;
use std::fs::File;
use std::io::Read;
use std::io::Write;
use std::path::Path;
use syn;

const BINDGEN_OUTPUT: &str = "src/ffi/sqlite3ext.rs";

fn main() {
    generate_ffi();
}

fn generate_ffi() {
    println!("cargo:rerun-if-env-changed=CARGO_FEATURE_STATIC");
    if let Some(_) = env::var_os("CARGO_FEATURE_STATIC") {
        // Nothing to do!
        return;
    } else {
        false
    };

    println!("cargo:rerun-if-changed={}", BINDGEN_OUTPUT);
    let mut file = File::open(format!("{}", BINDGEN_OUTPUT)).expect(BINDGEN_OUTPUT);
    let mut content = String::new();
    file.read_to_string(&mut content).unwrap();

    let ast = syn::parse_file(&content).unwrap();
    let ident = Ident::new("sqlite3_api_routines", Span::call_site());
    let api_routines: Vec<(&syn::Ident, &syn::TypeBareFn)> = ast
        .items
        .iter()
        .filter_map(|i| match i {
            syn::Item::Struct(s) => Some(s),
            _ => None,
        })
        .find(|s| s.ident == ident)
        .map(|s| match &s.fields {
            syn::Fields::Named(n) => Some(n.named.iter()),
            _ => None,
        })
        .unwrap_or(None)
        .expect("sqlite3_api_routines missing")
        .map(|field| {
            let name = field.ident.as_ref().expect("unnamed method");
            let method = extract_method(&field.ty).expect("invalid field");
            (name, method)
        })
        .collect();

    let init_lines: Vec<TokenStream> = api_routines
        .iter()
        .map(|(name, _)| {
            let sqlite3_name = format_ident!("sqlite3_{}", name);
            quote! {
                if let Some(x) = (*api).#name {
                    #sqlite3_name = x;
                }
            }
        })
        .collect();

    let methods: Vec<TokenStream> = api_routines
        .iter()
        .map(|(name, method)| {
            let unsafety = &method.unsafety;
            let abi = &method.abi;
            let args = &method.inputs;
            let varargs = &method.variadic;
            let ty = &method.output;
            let sqlite3_name = format_ident!("sqlite3_{}", name);
            quote! {
                pub static mut #sqlite3_name: #unsafety #abi fn (#args #varargs) #ty = unsafe { ::std::mem::transmute(unavailable as *mut ::std::os::raw::c_void) };
            }
        })
        .collect();

    let result = quote! {
        pub unsafe fn init_api_routines(api: *mut sqlite3_api_routines) {
            #(#init_lines)*
        }

        fn unavailable() { unreachable!() }

        #(#methods)*
    };

    let tokens = TokenStream::from(result);
    let src = format!("{}", tokens);
    let formatted = rustfmt(src).unwrap();
    let out_dir = env::var_os("OUT_DIR").unwrap();
    let dest_path = Path::new(&out_dir).join("sqlite3_api_routines.rs");
    fs::write(&dest_path, formatted).unwrap();
}

fn extract_method(ty: &syn::Type) -> Option<&syn::TypeBareFn> {
    match &ty {
        syn::Type::Path(tp) => tp.path.segments.last(),
        _ => None,
    }
    .map(|seg| match &seg.arguments {
        syn::PathArguments::AngleBracketed(args) => args.args.first(),
        _ => None,
    })?
    .map(|arg| match &arg {
        syn::GenericArgument::Type(t) => Some(t),
        _ => None,
    })?
    .map(|ty| match &ty {
        syn::Type::BareFn(r) => Some(r),
        _ => None,
    })?
}

fn rustfmt(input: String) -> Result<String, String> {
    let rustfmt =
        which::which("rustfmt").map_err(|e| format!("unable to locate rustfmt: {:?}", e))?;
    let mut rustfmt_child = std::process::Command::new(rustfmt)
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .spawn()
        .map_err(|e| format!("failed to spawn rustfmt: {:?}", e))?;
    let mut stdin = rustfmt_child.stdin.take().unwrap();

    ::std::thread::spawn(move || {
        stdin.write_all(input.as_bytes()).unwrap();
    });

    let output = rustfmt_child.wait_with_output().unwrap();
    if output.status.code() != Some(0) {
        return Err(format!("rustfmt exited with {:?}", output.status.code()));
    }
    String::from_utf8(output.stdout).map_err(|e| format!("rustfmt returned invalid utf-8: {:?}", e))
}
