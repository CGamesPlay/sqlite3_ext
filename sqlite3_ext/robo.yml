bindgen:
  summary: regenerate the sqlite3ext bindings
  command: |
    set -e
    which -s bindgen || cargo install bindgen
    bindgen src/ffi/sqlite3ext.h \
      --allowlist-file src/ffi/sqlite3.h \
      --allowlist-file src/ffi/sqlite3ext.h \
      --ignore-functions \
      --default-macro-constant-type signed \
      --raw-line "#![allow(non_snake_case)]" \
      --raw-line "#![allow(dead_code)]" \
      --raw-line "#![allow(non_camel_case_types)]" \
      -o src/ffi/sqlite3ext.rs
    echo Generated src/ffi/sqlite3ext.rs
