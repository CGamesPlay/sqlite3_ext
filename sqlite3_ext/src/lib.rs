use std::ffi::{c_void, CStr, CString};
pub use types::*;
pub use value::*;

pub mod ffi;
pub mod function;
pub mod types;
pub mod value;
pub mod vtab;

pub fn sqlite3_libversion() -> &'static str {
    let ret = unsafe { CStr::from_ptr(ffi::libversion()) };
    ret.to_str().expect("sqlite3_libversion")
}

/// Register the provided function to be called by each new database connection.
pub fn sqlite3_auto_extension(
    init: unsafe extern "C" fn(
        *mut ffi::sqlite3,
        *mut *mut std::os::raw::c_char,
        *mut ffi::sqlite3_api_routines,
    ) -> std::os::raw::c_int,
) -> Result<()> {
    let rc = unsafe {
        let init: unsafe extern "C" fn() = std::mem::transmute(init as *mut c_void);
        libsqlite3_sys::sqlite3_auto_extension(Some(init))
    };
    Error::from_sqlite(rc)
}

pub struct Connection {
    db: *mut ffi::sqlite3,
}

impl Connection {
    pub fn create_module<'vtab, T: vtab::VTab<'vtab> + 'vtab>(
        &self,
        name: &str,
        vtab: vtab::Module<'vtab, T>,
        aux: Option<T::Aux>,
    ) -> Result<()> {
        let name = CString::new(name).unwrap();
        let handle = Box::new(vtab::ModuleHandle { vtab, aux });
        let rc = unsafe {
            ffi::create_module_v2(
                self.db,
                name.as_ptr() as _,
                &handle.vtab.base,
                Box::into_raw(handle) as _,
                Some(drop_boxed::<vtab::ModuleHandle<T>>),
            )
        };
        match rc {
            ffi::SQLITE_OK => Ok(()),
            _ => Err(Error::Sqlite(rc)),
        }
    }
}

impl From<&rusqlite::Connection> for Connection {
    /// Convert a rusqlite::Connection to an sqlite3_ext::Connection.
    ///
    /// # Panics
    ///
    /// This method will panic if the sqlite3_ext API has not been initialized, see
    /// [sqlite3_auto_extension].
    fn from(conn: &rusqlite::Connection) -> Self {
        if !ffi::is_ready() {
            panic!("loadable extension not initialized");
        }
        unsafe {
            Connection {
                db: conn.handle() as _,
            }
        }
    }
}

impl From<*mut ffi::sqlite3> for Connection {
    fn from(db: *mut ffi::sqlite3) -> Connection {
        Connection { db }
    }
}

unsafe extern "C" fn drop_boxed<T>(data: *mut c_void) {
    let aux: Box<T> = Box::from_raw(data as _);
    std::mem::drop(aux);
}
