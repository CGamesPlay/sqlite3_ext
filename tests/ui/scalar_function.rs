use sqlite3_ext::{function::*, *};
use std::cell::Cell;

fn add_funcs(db: &Connection) -> Result<()> {
    let opts = FunctionOptions::default()
        .set_risk_level(RiskLevel::Innocuous)
        .set_n_args(0);
    let cell = Cell::new(42);
    db.create_scalar_function("drop_check", &opts, |c, _| c.set_result(cell.get()))?;
    Ok(())
}

fn main() {}
