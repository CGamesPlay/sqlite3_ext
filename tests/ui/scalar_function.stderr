error[E0373]: closure may outlive the current function, but it borrows `cell`, which is owned by the current function
 --> tests/ui/scalar_function.rs:9:52
  |
9 |     db.create_scalar_function("drop_check", &opts, |c, _| c.set_result(cell.get()))?;
  |                                                    ^^^^^^              ---- `cell` is borrowed here
  |                                                    |
  |                                                    may outlive borrowed value `cell`
  |
note: function requires argument type to outlive `'static`
 --> tests/ui/scalar_function.rs:9:5
  |
9 |     db.create_scalar_function("drop_check", &opts, |c, _| c.set_result(cell.get()))?;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: to force the closure to take ownership of `cell` (and any other referenced variables), use the `move` keyword
  |
9 |     db.create_scalar_function("drop_check", &opts, move |c, _| c.set_result(cell.get()))?;
  |                                                    ++++
